cpp/src/explainer/tree_shap.cu

```diff
@@ -0,0 +1,326 @@
+    feature_upper_bound = min(feature_upper_bound, other.feature_upper_bound);
+  }
+  static_assert(std::is_same<T, float>::value || std::is_same<T, double>::value,
+                "T must be a float or double");
+};
+
+enum class CondType : uint8_t {
+  kFloat, kDouble
+};
+
+class ExtractedPath {
+ public:
+  CondType cond_type;
+  int num_tree;
+  float global_bias;
+  tl::TaskType task_type;
+  tl::TaskParam task_param;
+  bool average_tree_output;
+
+  virtual CondType GetCondType() = 0;
+  virtual ~ExtractedPath() = default;
+
+  template <typename T>
+  static std::unique_ptr<ExtractedPath> Create();
+
+  template <typename Func, typename ...Args>
+  auto Dispatch(Func func, Args&& ...args);
+};
+
+template <typename T>
+class ExtractedPathImpl : public ExtractedPath {
+ public:
+  std::vector<gpu_treeshap::PathElement<MySplitCondition<T>>> paths;
+  ExtractedPathImpl() {
+    if (std::is_same<T, double>::value) {
+      cond_type = CondType::kDouble;
+    } else {
+      cond_type = CondType::kFloat;
+    }
+  }
+  virtual ~ExtractedPathImpl() = default;
+
+  CondType GetCondType() override {
+    return cond_type;
+  }
+};
+
+template <typename T>
+std::unique_ptr<ExtractedPath> ExtractedPath::Create() {
+  std::unique_ptr<ExtractedPath> model
+    = std::make_unique<ExtractedPathImpl<T>>();
+  return model;
+}
+
+template <typename Func, typename ...Args>
+auto
+ExtractedPath::Dispatch(Func func, Args&& ...args) {
+  switch (this->cond_type) {
+   case CondType::kDouble:
+    func(*dynamic_cast<ExtractedPathImpl<double>*>(this),
+         std::forward<Args>(args)...);
+    break;
+   case CondType::kFloat:
+   default:
+    func(*dynamic_cast<ExtractedPathImpl<float>*>(this),
+         std::forward<Args>(args)...);
+    break;
+  }
+}
+
+template <typename T>
+std::ostream& operator<<(
+    std::ostream& os,
+    const std::vector<gpu_treeshap::PathElement<MySplitCondition<T>>>& paths) {
+  std::vector<gpu_treeshap::PathElement<MySplitCondition<T>>> tmp(paths);
+  std::sort(tmp.begin(), tmp.end(),
+            [&](const gpu_treeshap::PathElement<MySplitCondition<T>>& a,
+                const gpu_treeshap::PathElement<MySplitCondition<T>>& b) {
+              if (a.path_idx < b.path_idx) return true;
+              if (b.path_idx < a.path_idx) return false;
+
+              if (a.feature_idx < b.feature_idx) return true;
+              if (b.feature_idx < a.feature_idx) return false;
+              return false;
+            });
+
+  for (auto i = 0ull; i < tmp.size(); i++) {
+    auto e = tmp[i];
+    if (i == 0 || e.path_idx != tmp[i - 1].path_idx) {
+      os << "path_idx:" << e.path_idx << ", leaf value:" << e.v;
+      os << "\n";
+    }
+    os << " (feature:" << e.feature_idx << ", pz:" << e.zero_fraction << ", ["
+       << e.split_condition.feature_lower_bound << "<=x<"
+       << e.split_condition.feature_upper_bound << "])";
+    os << "\n";
+  }
```

Comment:
Since we're just looping over `tmp` here, it suggests that we could replace the raw loop with an algorithm. How about something like the following (untested)?
```suggestion
  std::accumulate(std::begin(tmp), std::end(tmp), std::optional<std::size_t>{}, [&os](auto&& e, auto&& prev_path_idx) {
    if (!prev_path_idx || e.path_idx != *prev_path_idx) {
      os << "path_idx:" << e.path_idx << ", leaf value:" << e.v << "\n";
    }
    os << " (feature:" << e.feature_idx << ", pz:" << e.zero_fraction << ", ["
      << e.split_condition.feature_lower_bound << "<=x<"
      << e.split_condition.feature_upper_bound << "])"
      << "\n";
    return std::make_optional(e.path_idx);
  })
```